tests/microos/network_bonding.pm

```diff
@@ -11,45 +11,157 @@ use strict;
+
+    record_info("Testing Failover for Mode: $bond_mode", "NIC: $device");
+
     assert_script_run "ip link set dev $device down";
+    sleep $delay;
     script_run 'ip a';
-    # networking should be still good
-    assert_script_run 'ping -c1 -I bond0 conncheck.opensuse.org';
-    # bring back up device
+
+    # Validate bond mode and NIC statuses (the downed NIC should be "down")
+    validate_bond_mode_and_slaves($bond_name, $description, \@nics_status);
+
+    check_connectivity $bond_name;
+
     assert_script_run "ip link set dev $device up";
 }
 
-sub run {
-    my ($self) = @_;
-    select_console 'root-console';
-    my @devices;
-    # many device can share the same id, so we'll use hash keys to dedup
-    my %connections;
-    # remove existing NM-managed connections (except loopback).
-    foreach (split('\n', script_output 'nmcli -g DEVICE,UUID conn show --active')) {
-        next if /^lo:/;    # skip loopback device
-        my @item = split /:/;
-        push @devices, $item[0];
-        $connections{$item[1]} = 1;
+sub validate_bond_mode_and_slaves {
+    my ($bond_name, $description, $devices_ref) = @_;
+
+    assert_script_run "cat /proc/net/bonding/$bond_name | grep 'Mode:' | grep '$description'";
+
+    foreach my $device_info (@$devices_ref) {
+        my ($device, $status_up) = @$device_info;
+
+        my $expected_status = $status_up ? 'up' : 'down';
+
+        assert_script_run "grep -A 1 'Slave Interface: $device' /proc/net/bonding/$bond_name | grep 'MII Status: $expected_status'";
     }
-    assert_script_run "nmcli con delete '$_'" for keys %connections;
-    # create a new bonding interface and connect the two ethernet
-    assert_script_run "nmcli con add type bond ifname bond0 con-name bond0";
-    assert_script_run "nmcli con add type ethernet ifname $_ master bond0" for @devices;
-    # bring up bond interface
-    assert_script_run "nmcli con up bond0";
-    # reboot to ensure connection properly comes up at start
+}
+
+sub test_bonding_mode {
+    my ($self, $nics_ref, $miimon, $bond_name, $bond_mode, $description, $run_failover) = @_;
+    my @nics = @$nics_ref;
+
+    select_serial_terminal;
+
+    delete_existing_connections;
+
+    create_bond($bond_name, $bond_mode, $miimon);
+    add_devices_to_bond($bond_name, @nics);
+
+    assert_script_run "nmcli con up $bond_name";
+
     power_action('reboot', textmode => 1);
     $self->wait_boot;
-    select_console 'root-console';
-    # first connectivity check
-    assert_script_run 'ping -c1 -I bond0 conncheck.opensuse.org';
-    # check device failover
-    test_failover $_ for @devices;
+    select_serial_terminal;
+    check_connectivity $bond_name;
+
+    # Validate that all NICs are "up"
+    validate_bond_mode_and_slaves($bond_name, $description, [map { [$_, 1] } @nics]);
+
+    # Testing failover for each NIC
+    $run_failover && test_failover($bond_mode, $bond_name, $_, $description, \@nics) for @nics;
```

Comment:
I find this hard to read. It would be better to make it more explicit that you skip it only for `balance-xor`

Reply:
Plan is to get 802.3ad | balance-tlb | balance-alb at some point via MM, thus maybe balance-xor will be run as well

Reply:
I feel like that proves my point. I wouldn't include a rare exception into the data structure. Ofc, you are the implementor, hence it is your decision. The constant jumping between the lines of the code seems to me unnecessary